substrate/client/network/sync/src/blocks.rs

```diff
@@ -212,6 +212,30 @@ impl<B: BlockT> BlockCollection<B> {
 		ready
 	}
 
+	/// Returns the block header of the first block that is ready for importing.
+	/// `from` is the maximum block number for the start of the range that we are interested in.
+	/// The function will return None if the first block ready is higher than `from`.
+	pub fn first_ready_block_hdr(&self, from: NumberFor<B>) -> Option<B::Header> {
+		let mut prev = from;
+		for (&start, range_data) in &self.blocks {
+			if start > prev {
+				break
+			}
+
+			match range_data {
+				BlockRangeState::Complete(blocks) => {
+					let len = (blocks.len() as u32).into();
+					prev = start + len;
```

Comment:
This mirrors the logic in `ready_blocks()`. Looking at `ready_blocks()`, if a range where `start < from`,  and `range_data.blocks.len() == 0` (for whatever reason), it would break out of the loop in the next iteration.

Though I don't think we will ever have empty blocks in the Complete state, the way code is structured right now makes it possible.

Reply:
Should probably have a `NonEmptyVec<>` kind of abstraction here, to make it explicit

Reply:
OK, let's keep it as is for now. Could you add a comment that the code is structured like this because it should be consistent with `ready_blocks`?

Reply:
Added comment